blockchain/standalone/pow_test.go

```diff
@@ -353,3 +393,1700 @@ func TestCheckProofOfWork(t *testing.T) {
 		}
 	}
 }
+
+// TestCalcASERTDiff ensures the proof-of-work target difficulty calculation for
+// the algorithm defined by DCP0011 works as expected.
+func TestCalcASERTDiff(t *testing.T) {
+	t.Parallel()
+
+	// mockASERTParams houses the per-chain parameters used throughout the tests
+	// to mock networks.
+	type mockASERTParams struct {
+		powLimit           *big.Int
+		powLimitBits       uint32
+		targetSecsPerBlock int64
+		halfLifeSecs       int64
+	}
+
+	// mockMainNetParams houses the main network parameters to use throughout
+	// the tests.  They match the Decred mainnet params as of the time this
+	// comment was written.
+	one := big.NewInt(1)
+	mockMainNetParams := mockASERTParams{
+		powLimit:           new(big.Int).Sub(new(big.Int).Lsh(one, 224), one),
+		powLimitBits:       0x1d00ffff,
+		targetSecsPerBlock: 300,
+		halfLifeSecs:       43200,
+	}
+
+	// mockTestNetParams houses the test network parameters to use throughout
+	// the tests.  They match the Decred testnet params as of the time this
+	// comment was written.
+	mockTestNetParams := mockASERTParams{
+		powLimit:           new(big.Int).Sub(new(big.Int).Lsh(one, 232), one),
+		powLimitBits:       0x1e00ffff,
+		targetSecsPerBlock: 120,
+		halfLifeSecs:       720,
+	}
+
+	// mockSimNetParams houses the simulation network parameters to use
+	// throughout the tests.  They match the Decred simnet params as of the time
+	// this comment was written.
+	mockSimNetParams := mockASERTParams{
+		powLimit:           new(big.Int).Sub(new(big.Int).Lsh(one, 255), one),
+		powLimitBits:       0x207fffff,
+		targetSecsPerBlock: 1,
+		halfLifeSecs:       6,
+	}
+
+	// perInstanceTest is used to control each test instance by specifying the
+	// expected difficulty bits for given heights and timestamps.
+	type perInstanceTest struct {
+		height           uint64
+		timestamp        int64
+		expectedDiffBits uint32
+	}
+
+	tests := []struct {
+		name             string // test description
+		params           *mockASERTParams
+		startDiffBits    uint32 // starting difficulty bits for calculation
+		startHeight      int64  // starting height for calculation
+		startTime        int64  // starting time for calculation
+		perInstanceTests []perInstanceTest
+	}{{
```

Comment:
Since these are all reference test vectors from the DCP, I'm going to split them out in a separate JSON-formatted test vectors file and then update this code to parse that JSON and run the tests.

Having the test vectors external makes it much easier for other implementations to consume them and also helps ensure continuity between the DCP and these tests.

Reply:
Done.